### PR TITLE
feat(planner): unify execution of DML statements (MERGE, UPDATE, DELETE)

### PR DESCRIPTION
**Pull request by Dousir9**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

1. Unify execution of DML statements (MERGE, UPDATE, DELETE), the unified execution improves the performance of UPDATE and DELETE and avoids OOM caused by row id.
2. Support UDF expression for update set expression, there will be another PR to support UDF for other parts of DML later.

+ Part of https://github.com/datafuselabs/databend/issues/15292
+ Close https://github.com/datafuselabs/databend/issues/15806.

## Performance Test (Databend Cloud) (TPC-H SF10)

```sql
-- Update, predicate with subquery
update lineitem_60000000 set l_quantity = l_quantity + 1 where l_linenumber in (select l_linenumber from lineitem_100000);

-- Delete, predicate with subquery
delete from lineitem_60000000 where l_linenumber in (select l_linenumber from lineitem_100000);
```

### Databend Cloud (Small)
**Update Duration: 94.6s → 29.4s (321%)**
**Delete Duration: 80.7s → 16.3s (495%)**

### Databend Cloud (Medium)
**Update Duration: 70.2s → 24.6s (285%)**
**Delete Duration: 64.8s → 8.5s (762%)**

### Full Test Script
```sql
-- Small
-- 1. Prepare data
-- 1.1 TPC-H SF10
create or replace table lineitem_60000000 as select * from lineitem limit 60000000;
-- 1.2 Small table with 10w rows
create or replace table lineitem_100000 as select * from lineitem_60000000 limit 100000;

-- 2. Update test
-- 2.1 Sample predicate
-- main: 13s
-- pr:   13s
update lineitem_60000000 set l_quantity = l_quantity + 1 where l_linenumber >= 1 and l_linenumber < 7;

-- 2.2 Predicate with subquery
-- main: 94.6s
-- pr:   29.4s
update lineitem_60000000 set l_quantity = l_quantity + 1 where l_linenumber in (select l_linenumber from lineitem_100000);

-- 3. Delete test
-- 3.1 Predicate with subquery
-- main: 80.7s
-- pr:   16.3s
delete from lineitem_60000000 where l_linenumber in (select l_linenumber from lineitem_100000);


-- Medium
-- 1. Prepare data
-- 1.1 TPC-H SF10
create or replace table lineitem_60000000 as select * from lineitem limit 60000000;
-- 1.2 Small table with 10 w rows
create or replace table lineitem_100000 as select * from lineitem_60000000 limit 100000;

-- 2. Update test
-- 2.1 Sample predicate
-- main: 8s
-- pr:   8s
update lineitem_60000000 set l_quantity = l_quantity + 1 where l_linenumber >= 1 and l_linenumber < 7;

-- 2.2 Predicate with subquery
-- main: 70.2s
-- pr:   24.6s
update lineitem_60000000 set l_quantity = l_quantity + 1 where l_linenumber in (select l_linenumber from lineitem_100000);

-- 3. Delete test
-- 3.1 Predicate with subquery
-- main: 64.8s
-- pr:   8.5s
delete from lineitem_60000000 where l_linenumber in (select l_linenumber from lineitem_100000);
```

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [x] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16060)

### CODE REVIEW

**Comment by BohuTANG** [comment]
Will this refactor resolve the issue https://github.com/datafuselabs/databend/issues/15806?

**Comment by Dousir9** [comment]
> Will this refactor resolve the issue #15806?

Yeah.

**Review by xudong963 — COMMENTED** [review]
The overall migration framework and idea are as expected, most of the LGTM, but delete, update, insert execution details of the migration need to be further reviewed by the storage team. Good job!

**Comment by SkyFan2002** [comment]
LGTM. Thanks!

File: src/query/storages/fuse/src/operations/mutation_source.rs
[the remark this answers is not in the document]
Comment (Dousir9) [reply src/query/storages/fuse/src/operations/mutation_source.rs:1]: The data mutation of `MutationStrategy::Direct` occurs in `ColumnMutation`, and `mutation_source.rs` is used to read data.

**Review by zhyass — COMMENTED** [review]
It looks like the previous `try_eval_const` optimization for delete was lost.
https://github.com/datafuselabs/databend/blob/v1.2.592-nightly/src/query/service/src/interpreters/interpreter_delete.rs#L220-L229

**Review by zhyass — APPROVED** [review]

**Review by dantengsky — APPROVED** [review]

**Pull request merged**